### ISSUE TITLE
rpmlib-sys: popt.h causes bindgen to output invalid self-referential poptOption definition

### ISSUE THREAD

**Issue by tony-iqlusion** [iqlusion-io/crates#12]
NOTE: This is a tracking issue for blockers in **rpmlib-sys**. A separate issue will be opened for bindgen itself.

bindgen outputs invalid self-referential bindings for [/usr/include/popt.h](https://github.com/acquia/DevDesktopCommon/blob/master/bintools-win/msys/include/popt.h) and any headers that include it (see below for list):

```
error[E0428]: the name `poptOption` is defined multiple times
|
| pub struct poptOption {
| --------------------- previous definition of the type `poptOption` here
...
| pub type poptOption = *mut poptOption;
| ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `poptOption` redefined here
|
= note: `poptOption` must be defined only once in the type namespace of this module
```

It seems to be trying to create a type alias with the same name as the type it is aliasing:

```
pub type poptOption = *mut poptOption;
```

## Impacted headers/library bindings

This is a blocker for generating bindings for the following headers (and possibly more):

- rpmbuild.h
- rpmlib.h
- rpmsign.h
- rpmspec.h

**Comment by tony-iqlusion** [comment]
It seems the first step of filing a bindgen issue for this is creating a minimal reproduction of the issue:

https://github.com/rust-lang-nursery/rust-bindgen/issues/new

```C++
// Insert your minimal C or C++ header here.
//
// It should *NOT* have any `#include`s! Not all systems have the same header
// files, and therefore any `#include` harms reproducibility. Additionally,
// the test case isn't minimal since the included file almost assuredly
// contains things that aren't necessary to reproduce the bug, and makes
// tracking it down much more difficult.
//
// Use the `--dump-preprocessed-input` flag or the
// `bindgen::Builder::dump_preprocessed_input` method to make your test case
// standalone and without `#include`s, and then use C-Reduce to minimize it:
// https://github.com/rust-lang-nursery/rust-bindgen/blob/master/CONTRIBUTING.md#using-creduce-to-minimize-test-cases
```

I'll poke at this when I have time.

**Comment by tony-iqlusion** [comment]
The core problem is [popt.h] defines a typedef (for a pointer to a struct) with the same name as the struct, which is allowed in C, but not C++, and likewise bindgen generates invalid code for it:

https://github.com/acquia/DevDesktopCommon/blob/master/bintools-win/msys/include/popt.h#L205

```c
typedef struct poptOption * poptOption;
```

bindgen translates this to:

```
pub type poptOption = *mut poptOption;
```

...which does not compile.

There's gating for `#ifdef __cplusplus` to disable these typedefs, so one way to work around the problem is to have bindgen attempt to generate a C++ binding instead of a C one.

**Comment by tony-iqlusion** [comment]
This was fixed in #17 by having bindgen build a C++ binding instead of a C one

**Comment by tony-iqlusion** [comment]
Here's a relevant bindgen issue for the typedef name clash:

https://github.com/rust-lang-nursery/rust-bindgen/issues/1252
